main/src/main/scala/sbt/Keys.scala

```diff
@@ -268,6 +268,10 @@ object Keys {
   val outputStrategy = settingKey[Option[sbt.OutputStrategy]]("Selects how to log output when running a main class.").withRank(DSetting)
   val connectInput = settingKey[Boolean]("If true, connects standard input when running a main class forked.").withRank(CSetting)
   val javaHome = settingKey[Option[File]]("Selects the Java installation used for compiling and forking.  If None, uses the Java installation running the build.").withRank(ASetting)
+  val discoveredJavaHomes = settingKey[Map[JavaVersion, File]]("Discovered Java home directories")
+  val javaHomes = settingKey[Map[JavaVersion, File]]("The user-defined additional Java home directories")
+  val fullJavaHomes = taskKey[Map[JavaVersion, File]]("Combines discoveredJavaHomes and custom javaHomes.").withRank(CTask)
```

Comment:
`fullJavaHomes` is a task if someone wants to side effect. (Credit to Dale)

Reply:
I had to switch this back to a setting since carrying the state around aggregation is pain in the ass when implementing a command.